### PR TITLE
break(iot-dev): Redesign connection status change callback

### PR DESCRIPTION
**Pull request by timtay-microsoft**
Bundle all the fields into one context object so that we can add additional fields later as necessary

from:
```java
onStatusChanged(IotHubConnectionStatus status, IotHubConnectionStatusChangeReason statusChangeReason, Throwable throwable, Object callbackContext);
```

to:
```java
onStatusChanged(ConnectionStatusChangeContext connectionStatusChangeContext);
```

where ConnectionStatusChangeContext looks like:
```java
public class ConnectionStatusChangeContext
{
    @Getter
    IotHubConnectionStatus newStatus;

    @Getter
    IotHubConnectionStatusChangeReason newStatusReason;

    @Getter
    Throwable cause;

    @Getter
    Object callbackContext;
}

### CODE REVIEW

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ConnectionStatusChangeContext.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+public class ConnectionStatusChangeContext
+{
+    /**
+     * The new connection status of the client.
+     */
+    @Getter
+    IotHubConnectionStatus newStatus;
+
+    /**
+     * The reason why the sdk changed to this status.
```
Comment (drwill-ms) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ConnectionStatusChangeContext.java:20]:
```suggestion
     * The reason why status changed.
```

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ConnectionStatusChangeContext.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+     * The throwable that caused the change in status. May be null if there wasn't an associated throwable.
+     */
+    @Getter
+    Throwable cause;
+
+    /**
+     * The user provided context object that was set when setting the connection status change callback on the client.
+     */
+    @Getter
+    Object callbackContext;
```
Comment (drwill-ms) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ConnectionStatusChangeContext.java:35]: What might this be used for?

File: device/iot-device-samples/device-reconnection-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/IotHubConnectionStatusChangeLogger.java [hunk 3]
```diff
@@ -10,7 +11,12 @@
 @Slf4j
 public class IotHubConnectionStatusChangeLogger implements IotHubConnectionStatusChangeCallback {
     @Override
-    public void onStatusChanged(IotHubConnectionStatus status, IotHubConnectionStatusChangeReason statusChangeReason, Throwable throwable, Object callbackContext) {
+    public void onStatusChanged(ConnectionStatusChangeContext connectionStatusChangeContext) {
+        IotHubConnectionStatus status = connectionStatusChangeContext.getNewStatus();
+        IotHubConnectionStatusChangeReason statusChangeReason = connectionStatusChangeContext.getNewStatusReason();
+        Throwable throwable = connectionStatusChangeContext.getCause();
+
+
```
Comment (drwill-ms) [inline device/iot-device-samples/device-reconnection-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/IotHubConnectionStatusChangeLogger.java:19]: nit: multiple blank lines

**Review by drwill-ms — APPROVED** [review]

**Review by azabbasi — APPROVED** [review]

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ConnectionStatusChangeContext.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * The context surrounding a connection status change event for this client.
+ */
+@AllArgsConstructor
+public class ConnectionStatusChangeContext
+{
+    /**
+     * The new connection status of the client.
+     */
+    @Getter
+    IotHubConnectionStatus newStatus;
```
[the remark this answers is not in the document]
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ConnectionStatusChangeContext.java:17]: It wouldn't hurt, sure

**Pull request merged**